src/main/java/com/suse/saltstack/netapi/parser/JsonParser.java

```diff
@@ -122,4 +130,97 @@ public Stats deserialize(JsonElement jsonElement, Type type,
             }
         }
     }
+
+    /**
+     * Deserializer for Arguments class.
+     * Breaks the incoming arguments into args and kwargs parts
+     * and fills a new Arguments instance.
+     */
+    private class ArgumentsDeserializer implements JsonDeserializer<Arguments> {
+
+        private static final String KWARG_KEY = "__kwarg__";
+
+        @Override
+        public Arguments deserialize(JsonElement json, Type typeOfT,
+                JsonDeserializationContext context) throws JsonParseException {
+            Arguments result = new Arguments();
+
+            if (json != null && json.isJsonArray()) {
+                for (JsonElement jsonElement : json.getAsJsonArray()) {
+                    fillArgs(result, jsonElement);
+                }
+            }
+
+            return result;
+        }
+
+        /**
+         * Fills args/kwargs to given Arguments instance based on JSON data
+         * from jsonElement.
+         *
+         * @param result Arguments to be filled
+         * @param jsonElement input JSON data
+         */
+        private void fillArgs(Arguments result, JsonElement jsonElement) {
+            boolean isKwarg = isKwarg(jsonElement);
+            // we don't need __kwarg__ attribute:
+            // args/kwargs are stored to separate attributes of Arguments class
+            // so we don't need this information anymore
+            filterKwarg(jsonElement);
+
+            if (isKwarg) {
+                fillKwargsFromObject(result, jsonElement.getAsJsonObject());
+            } else {
+                result.getArgs().add(gson.fromJson(jsonElement, Object.class));
+            }
+        }
+
+        /**
+         * Fills kwargs from given jsonObject key/values to given Arguments instance.
+         *
+         * @param result Arguments to be filled
+         * @param jsonObject input json data
+         */
+        private void fillKwargsFromObject(Arguments result, JsonObject jsonObject) {
+            for (Map.Entry<String, JsonElement> kwItem : jsonObject.entrySet()) {
+                result.getKwargs().put(kwItem.getKey(),
+                        gson.fromJson(kwItem.getValue(),
+                        Object.class));
+            }
+        }
+
+        /**
+         * Checks whether element is kwarg or arg.
+         *
+         * @param jsonElement element to be tested
+         * @return true if element is kwarg
+         */
+        private boolean isKwarg(JsonElement jsonElement) {
+            if (!jsonElement.isJsonObject()) {
+                return false;
+            }
+
+            JsonObject asJsonObject = jsonElement.getAsJsonObject();
+
+            try {
```

Comment:
I agree with your comment from before that the boolean test with exception is ugly but there is a way around that `JsonPrimitive` has a `isBoolean` method. I had something like this in mind:
```java
private boolean isKwarg(JsonElement jsonElement) {
        if (jsonElement.isJsonObject()) {
            JsonObject jsonObject = jsonElement.getAsJsonObject();
            JsonElement kwarg =  jsonObject.get(KWARG_KEY);
        return kwarg != null
               && kwarg.isJsonPrimitive()
               && kwarg.getAsJsonPrimitive().isBoolean()
               && kwarg.getAsBoolean();
       } else {
            return false;
       }
}
```